zkevm-circuits/src/rlp_circuit_fsm.rs

```diff
@@ -0,0 +1,1439 @@
+                and::expr([
+                    meta.query_fixed(q_enabled, Rotation::cur()),
+                    is_not_padding.expr(),
+                ])
+            },
+            |meta| meta.query_advice(data_table.tx_id, Rotation::cur()),
+            |meta| meta.query_advice(data_table.tx_id, Rotation::next()),
+        );
+        let format_check = IsEqualChip::configure(
+            meta,
+            |meta| {
+                and::expr([
+                    meta.query_fixed(q_enabled, Rotation::cur()),
+                    is_not_padding.expr(),
+                ])
+            },
+            |meta| meta.query_advice(data_table.format, Rotation::cur()),
+            |meta| meta.query_advice(data_table.format, Rotation::next()),
+        );
+
+        // randomness values.
+        let evm_word_rand = challenges.evm_word();
+        let keccak_input_rand = challenges.keccak_input();
+
+        meta.create_gate("data table checks", |meta| {
+            let mut cb = BaseConstraintBuilder::default();
+
+            // if tx_id' == tx_id and format' == format then:
+            cb.condition(
+                and::expr([
+                    tx_id_check.is_equal_expression.expr(),
+                    format_check.is_equal_expression.expr(),
+                ]),
+                |cb| {
+                    // byte_idx' == byte_idx + 1
+                    cb.require_equal(
+                        "byte_idx increments",
+                        meta.query_advice(data_table.byte_idx, Rotation::next()),
+                        meta.query_advice(data_table.byte_idx, Rotation::cur()) + 1.expr(),
+                    );
+                    // byte_rev_idx' + 1 == byte_rev_idx
+                    cb.require_equal(
+                        "byte_rev_idx decrements",
+                        meta.query_advice(data_table.byte_rev_idx, Rotation::next()) + 1.expr(),
+                        meta.query_advice(data_table.byte_rev_idx, Rotation::cur()),
+                    );
+                    // bytes_rlc' == bytes_rlc * r + byte_value'
+                    cb.require_equal(
+                        "correct random linear combination of byte value",
+                        meta.query_advice(data_table.bytes_rlc, Rotation::next()),
+                        meta.query_advice(data_table.bytes_rlc, Rotation::cur())
+                            * keccak_input_rand.expr()
+                            + meta.query_advice(data_table.byte_value, Rotation::next()),
+                    );
+                },
+            );
+
+            // if tx_id' != tx_id then:
+            cb.condition(not::expr(tx_id_check.is_equal_expression.expr()), |cb| {
+                // tx_id' == tx_id + 1
+                cb.require_equal(
+                    "tx_id increments",
+                    meta.query_advice(data_table.tx_id, Rotation::cur()),
+                    meta.query_advice(data_table.tx_id, Rotation::next()),
+                );
```

Comment:
This might not be true. Conside the corner case in which `tx_id > 0 && tx_id' = 0`, the boolean `is_not_padding = true`, but clearly `tx_id != tx_id + 1`.

Reply:
Also, just noticed that the constraint itself is incorrect (there is no `+ 1.expr()`).

I know it won't satisfy the edge case. I was expecting a failure here, and then adding another condition to it, i.e. `next not padding` and a separate conditional check for `next is padding`